src/net/ifaddrs/mod.rs

```diff
@@ -0,0 +1,250 @@
+//! `ifaddrs` provides a safe interface for the system's network interface data.
+//!
+//! The `InterfaceAddrs` struct provides access to the system's network 
+//! interface data. You can either iterate over it or consume it and convert 
+//! it into an `InterfaceMap` (a `HashMap<String, Vec<InterfaceAddr>>`) for 
+//! more convenient access by interface name.
+//!
+//! # Examples
+//!
+//! You can access the basic information of the system in a few lines.
+//! The following program prints all the known addresses.
+//!
+//! ```
+//! use nix::net::ifaddrs::InterfaceAddrs;
+//!
+//! let addrs = InterfaceAddrs::query_system()
+//!     .expect("System has no network interfaces.");
+//!
+//! for addr in addrs {
+//!     println!("{}: {:?}", addr.name, addr.address);
+//! }
+//! ```
+//!
+//! The `IffFlags` struct provides access to info about the 
+//! state of an interface. This program prints the addresses of only
+//! interfaces which are up.
+//!
+//! ```
+//! use nix::net::ifaddrs::{InterfaceAddrs, iff_flags};
+//!
+//! let addrs = InterfaceAddrs::query_system()
+//!     .expect("System has no network interfaces.");
+//!
+//! for addr in addrs {
+//!     if addr.flags.contains(iff_flags::IFF_UP) {
+//!         println!("{}: {:?}", addr.name, addr.address);
+//!     }
+//! }
+//! ```
+//!
+//! You can convert the `InterfaceAddrs` struct into a `HashMap` easily.
+//! `InterfaceMap` is an alias for `HashMap<String, Vec<InterfaceAddr>>` for
+//! easier reference.
+//!
+//! ```
+//! use nix::net::ifaddrs::{InterfaceAddrs, InterfaceAddr, InterfaceMap};
+//! use std::collections::HashMap;
+//!
+//! let interfaces: InterfaceMap = 
+//!     InterfaceAddrs::query_system()
+//!     .expect("System has no network interfaces.")
+//!     .into(); // Convert to a hash map
+//!
+//! // Print all the addresses of the loopback interface
+//! if let Some(addrs) = interfaces.get("lo") {
+//!    println!("Loopback addresses:");
+//!    for addr in addrs {
+//!         println!("\t{:?}", addr);
+//!    }
+//! }
+//!     
+//! ```
+//!
+
+use libc;
+use std::net::{IpAddr};
+use std::ptr::null_mut;
+use std::ffi::CStr;
+use std::collections::HashMap;
+
+pub mod iff_flags;
+use self::iff_flags::IffFlags;
+
+mod sockaddr;
+use self::sockaddr::sockaddr_to_ipaddr;
+
+pub type InterfaceMap = HashMap<String, Vec<InterfaceAddr>>;
+
+/// Represents a handle into the operating system's knowledge about network 
+/// interfaces present on the system. Allows the user to iterate over 
+/// interface configurations.
+pub struct InterfaceAddrs {
+    inner: *mut libc::ifaddrs,
+    current: Option<&'static libc::ifaddrs>,
```

Comment:
It works just as well using either *mut or &'static; I was told on users.rustlang that it's more "rusty" to do it this way, but I'm happy to revert it if that's better style for the nix crate.

Reply:
I'm very skeptical.  The only reason why your patch compiles now is because you cast that lifetime within an `unsafe` block.  Could you please show me the post where somebody recommended using a static lifetime?

Reply:
Here's the post: https://users.rust-lang.org/t/code-review-for-my-first-libc-wrapper/12743/9?u=silverwingedseraph

As I said, I'm happy to switch it back.

Reply:
I think you should.